# Telco Churn Prediction


The Telco Churn Prediction Project aims to identify the factors that cause customers to stop using the telco service, also known as customer churn. To achieve this, the project analyses a large dataset of 7043 customer information, including demographics and service usage patterns. 

The ultimate goal is to create predictive models that can forecast customer behavior and predict which customers are at risk of churning. The project requires a thorough exploration of the Telco customer dataset to gain insights into the characteristics and behaviors associated with churn. This involves identifying patterns, understanding correlations between variables, and conducting a comprehensive analysis of features.

## Import Libraries and Load Data 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
Telco_customer_churn_status =  pd.read_excel('Telco_customer_churn_status.xlsx')
Telco_customer_churn_services =  pd.read_excel('Telco_customer_churn_services.xlsx')
Telco_customer_churn_population =  pd.read_excel('Telco_customer_churn_population.xlsx')
Telco_customer_churn_location =  pd.read_excel('Telco_customer_churn_location.xlsx')
Telco_customer_churn_demographics =  pd.read_excel('Telco_customer_churn_demographics.xlsx')

This data is part of the Base Samples that are available in IBM Cognos Analytics.
Downloaded from: 
[Telecommunications Industry Sample Data](https://accelerator.ca.analytics.ibm.com/bi/?perspective=authoring&pathRef=.public_folders%2FIBM%2BAccelerator%2BCatalog%2FContent%2FDAT00148&id=i9710CF25EF75468D95FFFC7D57D45204&objRef=i9710CF25EF75468D95FFFC7D57D45204&action=run&format=HTML&cmPropStr=%7B%22id%22%3A%22i9710CF25EF75468D95FFFC7D57D45204%22%2C%22type%22%3A%22reportView%22%2C%22defaultName%22%3A%22DAT00148%22%2C%22permissions%22%3A%5B%22execute%22%2C%22read%22%2C%22traverse%22%5D%7D)


The dataset contains information on a fictional California-based telecommunications company that provided internet and home phone services to 7,043 customers during Q3. It consists of five datasets containing information about demographics, location, population, services, and status.


| Dataframe         | Columns              |
|:---------------------|:--------------------------|
| Telco_customer_churn_status.xlsx              | Status ID, Customer ID, Count, Quarter, Satisfaction Score, Customer Status, Churn Label, Churn Value, Churn Score, CLTV, Churn Category, Churn Reason     |
| Telco_customer_churn_services.xlsx             | Service ID, Customer ID, Count, Quarter, Referred a Friend, Number of Referrals, Tenure in Months, Offer, Phone Service, Avg Monthly Long Distance Charges, Multiple Lines, Internet Service, Internet Type, Avg Monthly GB Download, Online Security, Online Backup, Device Protection Plan, Premium Tech Support, Streaming TV, Streaming Movies, Streaming Music, Unlimited Data, Contract, Paperless Billing, Payment Method, Monthly Charge, Total Charges, Total Refunds, Total Extra Data Charges, Total Long Distance Charges, Total Revenue |
| Telco_customer_churn_population.xlsx              | ID, Zip Code, Population    |
| Telco_customer_churn_location.xlsx              | Location ID, Customer ID, Count, Country, State, City, Zip Code, Lat Long, Latitude, Longitude    |
| Telco_customer_churn_demographics.xlsx             | Customer ID, Count, Gender, Age, Under 30, Senior Citizen, Married, Dependents, Number of Dependents    |

In [3]:
Telco_customer_churn_status.head()

Status ID Customer ID  Count Quarter  Satisfaction Score Customer Status  \
0  SWSORB1252  8779-QRDMV      1      Q3                   3         Churned   
1  SNAEQA8572  7495-OOKFY      1      Q3                   3         Churned   
2  LMBQNN3714  1658-BYGOY      1      Q3                   2         Churned   
3  VRZYZI9978  4598-XLKNJ      1      Q3                   2         Churned   
4  FDNAKX1688  4846-WHAFZ      1      Q3                   2         Churned   

  Churn Label  Churn Value  Churn Score  CLTV   Churn Category  \
0         Yes            1           91  5433       Competitor   
1         Yes            1           69  5302       Competitor   
2         Yes            1           81  3179       Competitor   
3         Yes            1           88  5337  Dissatisfaction   
4         Yes            1           67  2793            Price   

                   Churn Reason  
0  Competitor offered more data  
1  Competitor made better offer  
2  Competitor made better offer  
3     Limited range of services  
4            Extra data charges

In [4]:
Telco_customer_churn_services.head()

Service ID Customer ID  Count Quarter Referred a Friend  \
0  IJKDQVSWH3522  8779-QRDMV      1      Q3                No   
1  BFKMZJAIE2285  7495-OOKFY      1      Q3               Yes   
2  EIMVJQBMT7187  1658-BYGOY      1      Q3                No   
3  EROZQXDUU4979  4598-XLKNJ      1      Q3               Yes   
4  GEEYSJUHY6991  4846-WHAFZ      1      Q3               Yes   

   Number of Referrals  Tenure in Months    Offer Phone Service  \
0                    0                 1      NaN            No   
1                    1                 8  Offer E           Yes   
2                    0                18  Offer D           Yes   
3                    1                25  Offer C           Yes   
4                    1                37  Offer C           Yes   

   Avg Monthly Long Distance Charges  ... Unlimited Data        Contract  \
0                               0.00  ...             No  Month-to-Month   
1                              48.85  ...            Yes  Month-to-Month   
2                              11.33  ...            Yes  Month-to-Month   
3                              19.76  ...            Yes  Month-to-Month   
4                               6.33  ...            Yes  Month-to-Month   

  Paperless Billing   Payment Method Monthly Charge Total Charges  \
0               Yes  Bank Withdrawal          39.65         39.65   
1               Yes      Credit Card          80.65        633.30   
2               Yes  Bank Withdrawal          95.45       1752.55   
3               Yes  Bank Withdrawal          98.50       2514.50   
4               Yes  Bank Withdrawal          76.50       2868.15   

  Total Refunds Total Extra Data Charges Total Long Distance Charges  \
0          0.00                       20                        0.00   
1          0.00                        0                      390.80   
2         45.61                        0                      203.94   
3         13.43                        0                      494.00   
4          0.00                        0                      234.21   

  Total Revenue  
0         59.65  
1       1024.10  
2       1910.88  
3       2995.07  
4       3102.36  

[5 rows x 31 columns]

In [5]:
Telco_customer_churn_population.head()

ID  Zip Code  Population
0   1     90001       54492
1   2     90002       44586
2   3     90003       58198
3   4     90004       67852
4   5     90005       43019

In [6]:
Telco_customer_churn_location.head()

Location ID Customer ID  Count        Country       State         City  \
0  OXCZEW7397  8779-QRDMV      1  United States  California  Los Angeles   
1  FCCECI8494  7495-OOKFY      1  United States  California  Los Angeles   
2  HEHUQY7254  1658-BYGOY      1  United States  California  Los Angeles   
3  WIUHRF2613  4598-XLKNJ      1  United States  California    Inglewood   
4  CFEZBF4415  4846-WHAFZ      1  United States  California     Whittier   

   Zip Code                Lat Long   Latitude   Longitude  
0     90022   34.02381, -118.156582  34.023810 -118.156582  
1     90063  34.044271, -118.185237  34.044271 -118.185237  
2     90065  34.108833, -118.229715  34.108833 -118.229715  
3     90303  33.936291, -118.332639  33.936291 -118.332639  
4     90602  33.972119, -118.020188  33.972119 -118.020188

In [7]:
Telco_customer_churn_demographics.head()

Customer ID  Count  Gender  Age Under 30 Senior Citizen Married Dependents  \
0  8779-QRDMV      1    Male   78       No            Yes      No         No   
1  7495-OOKFY      1  Female   74       No            Yes     Yes        Yes   
2  1658-BYGOY      1    Male   71       No            Yes      No        Yes   
3  4598-XLKNJ      1  Female   78       No            Yes     Yes        Yes   
4  4846-WHAFZ      1  Female   80       No            Yes     Yes        Yes   

   Number of Dependents  
0                     0  
1                     1  
2                     3  
3                     1  
4                     1

## Data Wrangling

### Merge Data

In [8]:
df_merged = pd.merge(Telco_customer_churn_status.drop(columns=['Status ID','Count', 'Quarter']),
                     Telco_customer_churn_services.drop(columns=['Service ID','Count', 'Quarter']), 
                     on="Customer ID", how="inner")

In [9]:
df_merged_loc = pd.merge(df_merged, 
                         Telco_customer_churn_location.drop(columns=['Location ID','Count','Lat Long', 'Country','State']),
                         on="Customer ID", how="inner")
                         

In [10]:
df_merged_demo = pd.merge(df_merged_loc,
                          Telco_customer_churn_demographics.drop(columns=['Count','Under 30','Senior Citizen']),
                          on="Customer ID", how="inner")

In [11]:
df = pd.merge(df_merged_demo,
              Telco_customer_churn_population.drop(columns=['ID']),
              on="Zip Code", how="left")

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 46 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   object 
 1   Satisfaction Score                 7043 non-null   int64  
 2   Customer Status                    7043 non-null   object 
 3   Churn Label                        7043 non-null   object 
 4   Churn Value                        7043 non-null   int64  
 5   Churn Score                        7043 non-null   int64  
 6   CLTV                               7043 non-null   int64  
 7   Churn Category                     1869 non-null   object 
 8   Churn Reason                       1869 non-null   object 
 9   Referred a Friend                  7043 non-null   object 
 10  Number of Referrals                7043 non-null   int64  
 11  Tenure in Months                   7043 non-null   int64

In [13]:
df.to_csv('df.csv',index=False)

### Examine Missing Values

#### Churn Category and Churn Reason

In [14]:
df['Customer Status'].value_counts()

Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64

Out of the 7043 customers, 1869 churned, 4720 stayed, and 454 customers joined recently.

In [15]:
df['Churn Category'].unique()

array(['Competitor', 'Dissatisfaction', 'Price', 'Other', 'Attitude', nan],
      dtype=object)

In [16]:
df['Churn Reason'].unique()

array(['Competitor offered more data', 'Competitor made better offer',
       'Limited range of services', 'Extra data charges',
       'Competitor had better devices', "Don't know",
       'Service dissatisfaction',
       'Lack of affordable download/upload speed',
       'Product dissatisfaction', 'Long distance charges',
       'Poor expertise of online support', 'Attitude of support person',
       'Network reliability', 'Competitor offered higher download speeds',
       'Moved', 'Price too high', 'Attitude of service provider',
       'Poor expertise of phone support', 'Deceased', nan,
       'Lack of self-service on Website'], dtype=object)

The Churn Category and Churn Reason values are only available for customers who have churned. Otherwise, they are NaN which explains why there are only 1869 values in those columns.

#### Offer

In [17]:
df['Offer'].unique()

array([nan, 'Offer E', 'Offer D', 'Offer C', 'Offer B', 'Offer A'],
      dtype=object)

For applicable customers, the Offer field indicates the most recent marketing offer that was accepted. The following values are possible: 'None', 'Offer A', 'Offer B', 'Offer C', 'Offer D', and 'Offer E'. If no offer has been accepted, the field will display 'None'. In case of missing data, please fill it with 'No Offer'.

In [18]:
df['Offer'] = df['Offer'].fillna('No Offer')

#### Internet Type

Company offers two types of services: phone and internet. Create a new column for subscription type, which indicates if customers subscribe to phone service only, internet service only, or both.


In [19]:
df['Subscription Type'] = 'Both Services'
df.loc[df['Phone Service'] == 'No', 'Subscription Type'] = 'Internet Only'
df.loc[df['Internet Service'] == 'No', 'Subscription Type'] = 'Phone Only'

In [20]:
df['Subscription Type'].value_counts()

Subscription Type
Both Services    4835
Phone Only       1526
Internet Only     682
Name: count, dtype: int64

In [21]:
df['Internet Type'].unique()

array(['DSL', 'Fiber Optic', 'Cable', nan], dtype=object)

The missing value for Internet Type can be explained by the total number of internet subscribers.



In [22]:
df.to_csv('df.csv',index=False)